pkg/daemon/controller.go

```diff
@@ -318,31 +292,51 @@ func (c *Controller) initProviderNetwork(pn *kubeovnv1.ProviderNetwork, node *v1
 		klog.Errorf("failed to patch node %s: %v", node.Name, err)
 		return err
 	}
-
+	c.recordProviderNetworkErr(pn.Name, "")
 	return nil
 }
 
-func (c *Controller) updateProviderNetworkStatusForNodeDeletion(pn *kubeovnv1.ProviderNetwork, node string) error {
-	var needUpdate bool
-	if util.ContainsString(pn.Status.ReadyNodes, node) {
-		pn.Status.ReadyNodes = util.RemoveString(pn.Status.ReadyNodes, node)
-		needUpdate = true
-	}
-	if pn.Status.RemoveNodeConditions(node) {
-		needUpdate = true
+func (c *Controller) recordProviderNetworkErr(providerNetwork string, errMsg string) {
+	var currentPod *v1.Pod
+	var err error
+	if c.localPodName == "" {
+		pods, err := c.config.KubeClient.CoreV1().Pods("").List(context.Background(), metav1.ListOptions{
+			LabelSelector: "app=kube-ovn-cni",
+			FieldSelector: fmt.Sprintf("spec.nodeName=%s", c.config.NodeName),
+		})
+		if err != nil {
+			klog.Errorf("failed to list pod: %v", err)
+			return
+		}
+		for _, pod := range pods.Items {
+			if pod.Spec.NodeName == c.config.NodeName && pod.Status.Phase == v1.PodRunning {
+				c.localPodName = pod.Name
+				c.localNamespace = pod.Namespace
+				currentPod = &pod
+				break
+			}
+		}
+	} else {
+		if currentPod, err = c.config.KubeClient.CoreV1().Pods(c.localNamespace).Get(context.Background(), c.localPodName, metav1.GetOptions{}); err != nil {
+			klog.Errorf("failed to get pod %s, %v", c.localPodName, err)
+			return
+		}
 	}
 
-	if !needUpdate {
-		return nil
+	if currentPod.Annotations == nil {
+		currentPod.Annotations = make(map[string]string)
 	}
-
-	_, err := c.config.KubeOvnClient.KubeovnV1().ProviderNetworks().UpdateStatus(context.Background(), pn, metav1.UpdateOptions{})
-	if err != nil {
-		klog.Errorf("failed to update status of provider network %s: %v", pn.Name, err)
-		return err
+	if errMsg == "" {
+		delete(currentPod.Annotations, fmt.Sprintf(util.ProviderNetworkErrMessageTemplate, providerNetwork))
+	} else {
+		currentPod.Annotations[fmt.Sprintf(util.ProviderNetworkErrMessageTemplate, providerNetwork)] = errMsg
 	}
 
-	return nil
+	patchPayload := []byte(fmt.Sprintf(`[{ "op": "replace", "path": "/metadata/annotations", "value": %s }]`, currentPod.Annotations))
+	if _, err = c.config.KubeClient.CoreV1().Pods(c.localNamespace).Patch(context.Background(), c.localPodName, types.JSONPatchType, patchPayload, metav1.PatchOptions{}); err != nil {
+		klog.Errorf("failed to patch pod %s: %v", c.localPodName, err)
+		return
+	}
```

Comment:
We should use the strategic merge patch method.

https://github.com/kubeovn/kube-ovn/pull/1480

Reply:
We should also check whether the annotations are changed before applying the patch.